In [38]:
import os
os.getcwd()

'C:\\Users\\MS00420757\\Documents\\Python Scripts\\Automobile Hands on'

In [39]:
your_local_path = os.getcwd()

###  Import required modules

In [40]:
import pandas as pd               
import numpy as np
import pickle

from sklearn.model_selection import train_test_split   #splitting data

from sklearn.linear_model import LinearRegression         #linear regression
from sklearn.metrics.regression import mean_squared_error #error metrics
from sklearn.metrics import mean_absolute_error

import seaborn as sns                       #visualisation
import matplotlib.pyplot as plt             #visualisation

%matplotlib inline     
sns.set(color_codes=True)

###  Read data

In [41]:
automobile_data = pd.read_csv('Automobile_data.csv')

### Inspect and explore 

In [43]:
automobile_data.head()                               # check the head of the data

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3                 ?  alfa-romero       gas        std          two   
1          3                 ?  alfa-romero       gas        std          two   
2          1                 ?  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...    engine-size  \
0  convertible          rwd           front        88.6  ...            130   
1  convertible          rwd           front        88.6  ...            130   
2    hatchback          rwd           front        94.5  ...            152   
3        sedan          fwd           front        99.8  ...            109   
4        sedan          4wd           front        99.4  ...            136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

In [44]:
automobile_data.describe()                        # summary statistucs

symboling  wheel-base      length       width      height  \
count  205.000000  205.000000  205.000000  205.000000  205.000000   
mean     0.834146   98.756585  174.049268   65.907805   53.724878   
std      1.245307    6.021776   12.337289    2.145204    2.443522   
min     -2.000000   86.600000  141.100000   60.300000   47.800000   
25%      0.000000   94.500000  166.300000   64.100000   52.000000   
50%      1.000000   97.000000  173.200000   65.500000   54.100000   
75%      2.000000  102.400000  183.100000   66.900000   55.500000   
max      3.000000  120.900000  208.100000   72.300000   59.800000   

       curb-weight  engine-size  compression-ratio    city-mpg  highway-mpg  
count   205.000000   205.000000         205.000000  205.000000   205.000000  
mean   2555.565854   126.907317          10.142537   25.219512    30.751220  
std     520.680204    41.642693           3.972040    6.542142     6.886443  
min    1488.000000    61.000000           7.000000   13.000000    16.000000  
25%    2145.000000    97.000000           8.600000   19.000000    25.000000  
50%    2414.000000   120.000000           9.000000   24.000000    30.000000  
75%    2935.000000   141.000000           9.400000   30.000000    34.000000  
max    4066.000000   326.000000          23.000000   49.000000    54.000000

In [45]:
automobile_data.info()                         

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
symboling            205 non-null int64
normalized-losses    205 non-null object
make                 205 non-null object
fuel-type            205 non-null object
aspiration           205 non-null object
num-of-doors         205 non-null object
body-style           205 non-null object
drive-wheels         205 non-null object
engine-location      205 non-null object
wheel-base           205 non-null float64
length               205 non-null float64
width                205 non-null float64
height               205 non-null float64
curb-weight          205 non-null int64
engine-type          205 non-null object
num-of-cylinders     205 non-null object
engine-size          205 non-null int64
fuel-system          205 non-null object
bore                 205 non-null object
stroke               205 non-null object
compression-ratio    205 non-null float64
horsepower           205 non-nul

### Data cleaning

#### Replace unwanted symbol "?" with NaN 

In [46]:
automobile_data.replace('?', np.nan, inplace=True)

#### Count the number of missing values in the dataframe

In [47]:
# count the number of NaN values in each column
print(automobile_data.isnull().sum())

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64


#### Remove columns that are 90% empty

In [48]:
## remove columns that are 90% empty


thresh = len(automobile_data) * .1
automobile_data.dropna(thresh = thresh, axis = 1, inplace = True)

Here axis refers to a direction along which aggregation will take place in the matrix, axis=0 refers to row wise aggregation whereas axis=1 refers to column wise aggregation
When inplace=True is passed, the data is renamed in place .i.e in the same cell itself.

In [49]:
# count the number of NaN values in each column
print(automobile_data.isnull().sum())

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-location       0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
engine-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64


### Data Imputation

Fill the null value cells using appropriate values of the particular column using aggregation functions such as mean, median or mode. 

In [50]:
## Define a function impute_median
def impute_median(series):
    return series.fillna(series.median())

#automobile_data['num-of-doors']=automobile_data['num-of-doors'].transform(impute_median)
automobile_data.bore=automobile_data['bore'].transform(impute_median)
automobile_data.stroke=automobile_data['stroke'].transform(impute_median)
automobile_data.horsepower=automobile_data['horsepower'].transform(impute_median)
automobile_data.price=automobile_data['price'].transform(impute_median)


In [51]:
automobile_data['num-of-doors'].fillna(str(automobile_data['num-of-doors'].mode().values[0]),inplace=True)
automobile_data['peak-rpm'].fillna(str(automobile_data['peak-rpm'].mode().values[0]),inplace=True)
automobile_data['normalized-losses'].fillna(str(automobile_data['normalized-losses'].mode().values[0]),inplace=True)

In [52]:
# count the number of NaN values in each column
print(automobile_data.isnull().sum())

symboling            0
normalized-losses    0
make                 0
fuel-type            0
aspiration           0
num-of-doors         0
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 0
stroke               0
compression-ratio    0
horsepower           0
peak-rpm             0
city-mpg             0
highway-mpg          0
price                0
dtype: int64


In [53]:
automobile_data.head()

symboling normalized-losses         make fuel-type aspiration num-of-doors  \
0          3               161  alfa-romero       gas        std          two   
1          3               161  alfa-romero       gas        std          two   
2          1               161  alfa-romero       gas        std          two   
3          2               164         audi       gas        std         four   
4          2               164         audi       gas        std         four   

    body-style drive-wheels engine-location  wheel-base  ...    engine-size  \
0  convertible          rwd           front        88.6  ...            130   
1  convertible          rwd           front        88.6  ...            130   
2    hatchback          rwd           front        94.5  ...            152   
3        sedan          fwd           front        99.8  ...            109   
4        sedan          4wd           front        99.4  ...            136   

   fuel-system  bore  stroke compression-ratio horsepower  peak-rpm city-mpg  \
0         mpfi  3.47    2.68               9.0        111      5000       21   
1         mpfi  3.47    2.68               9.0        111      5000       21   
2         mpfi  2.68    3.47               9.0        154      5000       19   
3         mpfi  3.19     3.4              10.0        102      5500       24   
4         mpfi  3.19     3.4               8.0        115      5500       18   

  highway-mpg  price  
0          27  13495  
1          27  16500  
2          26  16500  
3          30  13950  
4          22  17450  

[5 rows x 26 columns]

###  Prepare the data

In [142]:
X = automobile_data.drop(['price'],axis=1)
X = automobile_data[['engine-size','city-mpg']]
y = automobile_data.price

X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=12345)

### Train 

Let us train the housing dataset using a **linear regression** model. **Linear Regression** is a form of regression that models **linear relationship** between **dependent** and **independent** variables.

Linear regression equation is given as follows:

$y = \beta_0 + \beta_1x_1 + .....+\beta_nx_n $

- $y$ is the response
- $\beta_0$ is the intercept
- $\beta_1$ is the coefficient for $x_1$ (the first feature)
- $\beta_n$ is the coefficient for $x_n$ (the second feature)

In [143]:
# initialize the model
lr= LinearRegression()

# fit the model
model_fit=lr.fit(X_train,y_train)


In [144]:
# print the coefficients
print (model_fit.intercept_)
print (model_fit.coef_)

597.4537790458617
[ 140.12847385 -214.9315153 ]


Interpreting the **intercept** ($\beta_0$):

- It is the value of $y$ when $x$=0.

Interpreting the coefficients ($\beta_1$,$\beta_2$):

- Holding all other features fixed, a 1 unit increase in **engine size** is associated with a **price increase of 140**.
- Holding all other features fixed, a 1 unit increase in **is associated with a** **price decrease of 214**.


In [145]:
sale_price=model_fit.predict(np.array([[130,1]]))
print("The sale price is:",sale_price)

The sale price is: [18599.22386446]


Linear regression equation is given as follows:

$y = \beta_0 + \beta_1x_1 + \beta_2x_2 $

In [146]:
y=model_fit.intercept_+(130*model_fit.coef_[0]+1*model_fit.coef_[1])
print("The sale price is:",y)

The sale price is: 18599.223864459986


#### Train Evaluation

In [147]:
#predict on test data
train_pred = model_fit.predict(X_train)

#mean squared error
mse=mean_squared_error(y_train,train_pred)

#root mean squared error
print('train rmse: {}'.format(np.sqrt(mse)))

#mean absolute error
mae=mean_absolute_error(y_train,train_pred)
print('train mae: {}'.format(mae))

train rmse: 3790.921768293673
train mae: 2746.082663590994


### Test

In [148]:
#predict on test data
test_pred = model_fit.predict(X_test)

#### Test Evaluation

In [149]:
#mean squared error
mse=mean_squared_error(y_test,test_pred)

#root mean squared error
print('test rmse: {}'.format(np.sqrt(mse)))

#mean absolute error
mae=mean_absolute_error(y_test,test_pred)
print('test mae: {}'.format(mae))

test rmse: 4040.46058916698
test mae: 2722.1984316838925
